src/troute-routing/troute/routing/diffusive_utils_v02.py

```diff
@@ -0,0 +1,1187 @@
+                        size_bathy_g[seg, frj] = nstations
+                        
+                        # populate cross section x, z and mannings n arrays
+                        x_bathy_g[0:nstations, seg, frj]    = topobathy_bytw.loc[seg_idx].xid_d
+                        z_bathy_g[0:nstations, seg, frj]    = topobathy_bytw.loc[seg_idx].z
+                        mann_bathy_g[0:nstations, seg, frj] = topobathy_bytw.loc[seg_idx].n
+                        
+                        # if terminal node of the network, then adjust the cross section z data using
+                        # channel slope and length data
+                        if segID == dbfksegID:
+                            So = param_df.loc[seg_idx].s0
+                            dx = param_df.loc[seg_idx].dx
+                            z_bathy_g[0:nstations, seg, frj] = z_bathy_g[0:nstations, seg, frj] - So * dx   
+   
+    else: 
+        #if the bathy dataframe is empty, then pass out empty arrays
+        x_bathy_g    = np.array([]).reshape(0,0,0)
+        z_bathy_g    = np.array([]).reshape(0,0,0)
+        mann_bathy_g = np.array([]).reshape(0,0,0)
+        size_bathy_g = np.array([], dtype = 'i4').reshape(0,0)
+        mxnbathy_g   = int(0)
+    
+    return x_bathy_g, z_bathy_g, mann_bathy_g, size_bathy_g, mxnbathy_g
+
+def fp_da_map(
+    mx_jorder,
+    ordered_reaches,
+    usgs_df,
+    nrch_g,
+    t0,
+    nsteps,
+    dt_da_g,
+    t0_g,
+    tfin_g):
+    """
+    Data assimilatoin data mapping between Python and Fortran
+
+    Parameters
+    ----------
+    mx_jorder       -- (int) maximum network reach order
+    ordered_reaches -- (dict) reaches and reach metadata by junction order
+    usgs_df         -- (DataFrame) usgs streamflow data interpolated at dt time steps
+    t0              -- (datetime) initial date
+    nsteps          -- (int) number of simulation time steps
+    dt_da_g         -- (int) numer of Data Assimilation timesteps
+    t0_g            -- (float) diffusive model's initial simulation time [hr] (by default, zero)
+    tfin_g          -- (float) diffusive model's final simulation time [hr] 
+
+    Returns
+    -------
+    nts_da_g        -- (int) number of DA timesteps
+    usgs_da_g       -- (float) usgs oberved streamflow data [cms]
+    usgs_da_reach_g -- (int) indices of stream reaches where DA is applied    
+    """ 
+    nts_da_g    = int((tfin_g - t0_g) * 3600.0 / dt_da_g) + 1  # include initial time 0 to the final time    
+    usgs_da_g   = -4444.0*np.ones((nts_da_g, nrch_g))
+    usgs_da_reach_g = np.zeros(nrch_g, dtype='i4')
+    
+    if not usgs_df.empty:    
+        dt_timeslice = timedelta(minutes=dt_da_g/60.0)
+        tfin         = t0 + dt_timeslice*nsteps
+        timestamps   = pd.date_range(t0, tfin, freq=dt_timeslice)
+    
+        usgs_df_complete = usgs_df.replace(np.nan, -4444.0)
+    
+        for i in range(len(timestamps)):
```

Comment:
Instead of for loop because we keep adding column to dataframe it slow it down and give warning. This should solve the issue
```python
        missing_timestamps = [ts for ts in timestamps if ts not in usgs_df.columns]
    
  
        missing_data = pd.DataFrame(-4444.0*np.ones((len(usgs_df), len(missing_timestamps))), 
                                    columns=missing_timestamps, 
                                    index=usgs_df.index)
        
       
        usgs_df_complete = pd.concat([usgs_df_complete, missing_data], axis=1)

Reply:
@AminTorabi-NOAA @shorvath-noaa These lines replace existing lines 537~539:
https://github.com/kumdonoaa/t-route/blob/4c56cc615b49f6cc75ede3c6b8879689940bee9a/src/troute-routing/troute/routing/diffusive_utils_v02.py#L537-L539

Reply:
Yes. As Sean suggested you can add `usgs_df_complete = usgs_df_complete[timestamps]` at the end of those lines too